src/models/adVAE.jl

```diff
@@ -0,0 +1,187 @@
+        ConvDecoder(isize, 1, in_ch, nf, extra_layers) 
+    )
+
+    return adVAE(encoder, generator, transformer)
+end
+
+Flux.@functor adVAE
+
+function (advae::adVAE)(x)
+    μ, Σ = advae.encoder(x)
+    z = μ + Σ * randn(Float32)
+    xᵣ = advae.generator(z)
+end
+
+kl_divergence(μ, Σ) = 0.5 * sum(1 .+ log.(Σ.^2) - μ.^2  - Σ.^2) # actually it is -kl_divergence
+kl_divergence(μ₁, Σ₁, μ₂, Σ₂) = sum(log.(Σ₂ ./ Σ₁) + (Σ₁.^2 + (μ₁ - μ₂).^2) ./ (2*Σ₂.^2) .- 0.5)
+
+function loss(advae::adVAE, x; γ=1e-3, λ=1e-2, mx=1, mz=1)
+    μ, Σ = advae.encoder(x)
+    μₜ, Σₜ = advae.transformer(cat(μ, Σ, dims=1))
+    
+    z = μ + Σ * randn(Float32)
+    zₜ = μₜ + Σₜ * randn(Float32)
+    
+    xᵣ = advae.generator(z)
+    xₜᵣ = advae.generator(zₜ)
+    
+    μᵣ, Σᵣ = advae.encoder(xᵣ)
+    μₜᵣ, Σₜᵣ = advae.encoder(xₜᵣ)
+
+    # 𝓛 for generator => 𝓛 = 𝓛_z + 𝓛_zₜ 
+    𝓛_z = Flux.Losses.mse(x, xᵣ) .+ γ * kl_divergence(μᵣ, Σᵣ)
+    𝓛_zₜ = max(0, (mx - Flux.Losses.mse(xᵣ, xₜᵣ))) + γ * max(0, (mz - kl_divergence(μₜᵣ, Σₜᵣ)))
+    𝓛 = 𝓛_z + 𝓛_zₜ
+    # 𝓛ₜ for transformer 
+    𝓛ₜ = kl_divergence(μ, Σ, μₜ, Σₜ)
+    # 𝓛ₑ for encoder
+    𝓛ₑ = Flux.Losses.mse(x, xᵣ) .+ γ * kl_divergence(μ, Σ) 
+        + γ * max(0, (mz - kl_divergence(μᵣ, Σᵣ)))
+        + γ * max(0, (mz - kl_divergence(μₜᵣ, Σₜᵣ)))
+
+    return 𝓛 + λ*𝓛ₜ, 𝓛ₑ, 𝓛, 𝓛ₜ
+end
+
+
+function anomaly_score(advae::adVAE, real_input; L=100, dims=3, batch_size=64)
+    real_input = Flux.Data.DataLoader(real_input, batchsize=batch_size)
+    testmode!(advae)
+    advae = advae |> gpu
+    output = Array{Float32}([])
+    for x in real_input
+        X = Array{Float32}(undef, size(x)[end], 1)
+        μ, Σ = advae.encoder(x)
+        for l=1:L
+            z = μ + Σ * randn(Float32)
+            xᵣ = advae.generator(z)
+            rec_loss = vec(Flux.mse(latent_i, latent_o, agg=x->mean(x, dims=dims))) |> cpu # loss per batch
+            X = cat(X, rec_loss, dims=2)
+        end
+        output = cat(output, Flux.mean(X, dims=2), dims=1)
+    end
+    testmode!(advae, false)
+    return output
+end
+
+
+function StatsBase.fit!(advae::adVAE, data, params)
+	# prepare batches & loaders
+	train_loader, val_loader = prepare_dataloaders(data, params)
+	# training info logger
+	history = MVHistory()
+	# prepare for early stopping
+	best_adVAE = deepcopy(advae)
+	patience = params.patience
+	best_val_loss = 1e10
+	val_batches = length(val_loader)
+
+	# ADAMW(η = 0.001, β = (0.9, 0.999), decay = 0) = Optimiser(ADAM(η, β), WeightDecay(decay))
+	opt = haskey(params, :decay) ? ADAMW(params.lr, (0.9, 0.999), params.decay) : ADAM(params.lr)
+
+	ps_step1 = Flux.params(advae.generator, advae.transformer)
+	ps_step2 = Flux.params(advae.encoder)
+
+
+	progress = Progress(length(train_loader))
+	for (iter, X) in enumerate(train_loader)
+		# training step 1
+		loss_1, back = Flux.pullback(ps_step1) do
+            loss(advae|>gpu , getobs(X)|>gpu, γ=params.gamma, λ=params.lambda, mx=params.mx, mz=params.mz)
```

Comment:
What if you train on tabular data? Than this does not work. Move the model to gpu outside the fit function. Also, I would create a separate loss for tabular and image data, where the other is going to move them on gpu, and set it as an argument of the fit! function. See the vae.jl run scripts for tabular and image data.

Reply:
it works on tabular data. There is no problem. 
Moreover when you dont have gpu on your system, it automaticly works only on cpu with no error or warning.
But if you want i will chagne it.